File: 19EulerEcuacionesDif/euler.py
import numpy as np
import sympy as sy
from sympy import symbols
import matplotlib.pyplot as plt

def euler (f,x_0,y_0,a,b,numeroPuntos):
    h = (b-a)/(numeroPuntos - 1)

    vectorX = obtenerVectorX(h,numeroPuntos,a,b)
    
    XY = vectoresXY(f,x_0,y_0,h,vectorX)
    print('x',XY[0])
    print('y',XY[1])
    plt.plot(XY[0],XY[1])
    plt.show()
def obtenerVectorX(h,numeroPuntos,a,b):
    vectorX = np.zeros([numeroPuntos])
    contador = 0

    while contador <= numeroPuntos-1:
        if contador == 0:
            vectorX[contador] = a
            contador += 1
        elif contador == numeroPuntos:
            vectorX[contador] = b
            contador += 1
        else:
            a = a + h
            vectorX[contador] = a
            contador += 1
    
    return vectorX
    


def vectoresXY(f,x_0,y_0,h,vectorX):
    x,y = symbols('x y')
    contador = 0
    m = len(vectorX)
    Y = np.zeros([m])
    
    Y[contador] = y_0
    f = sy.sympify(f)
    contador += 1
    while contador != m:
        
        funcion = f.evalf(subs={x:vectorX[contador-1],y:Y[contador - 1]})

        Y[contador] = Y[contador - 1] + h*(funcion)

        contador += 1
    
    return [vectorX, Y]



euler('y - x^2 + 1',0,0.5,0,2,11)

File: 19EulerEcuacionesDif/euler.m
function [X Y] = euler(f,x_0,y_0,a,b,numeroPuntos)
  pkg load symbolic
  h = (b-a)/(numeroPuntos - 1);
  
  vectorX = obtenerVectorX(h,numeroPuntos,a,b);
  
  [X Y] = vectoresXY(f,x_0,y_0,h,vectorX)
  plot(X ,Y)
  return
endfunction

function vectorX = obtenerVectorX(h,numeroPuntos,a,b)
  vectorX = zeros(1,numeroPuntos);
  contador = 1;
  while contador <= numeroPuntos
    if contador == 1
      vectorX(contador) = a;
      contador = contador + 1;
    elseif contador == numeroPuntos;
      vectorX(contador) = b;
      contador = contador + 1;
    else
      a = a+h;
      vectorX(contador) = a;
      contador = contador + 1; 
    endif
  
  endwhile
  return 
endfunction

function [X Y] = vectoresXY(f,x_0,y_0,h,vectorX)
  syms x y;
  m = length(vectorX);
  contador = 1;
  
  Y = zeros(1,vectorX);
  Y(contador) = y_0;
  contador = 2;
  
  while contador != m + 1
    x = vectorX(contador - 1);
    y = Y(contador - 1);
    funcion = eval(f);
    Y(contador) = Y(contador-1) + h*(funcion);
    
    contador = contador + 1;
    
  endwhile
  X = vectorX;
  return
  
endfunction
